orc8r/cloud/go/services/bootstrapper/servicers/registration/registration.go

```diff
@@ -114,3 +119,43 @@ func makeErr(errString string) *protos.RegisterResponse {
 	}
 	return errRes
 }
+
+// createGatewayDevice creates the gateway device model
+func createGatewayDevice(hwID *protos.AccessGatewayID, challengeKey *protos.ChallengeKey) *models.GatewayDevice {
+	challengeKeyBase64 := strfmt.Base64(challengeKey.Key)
+	return &models.GatewayDevice{
+		HardwareID: hwID.Id,
+		Key:        &models.ChallengeKey{KeyType: challengeKey.KeyType.String(), Key: &challengeKeyBase64},
+	}
+}
+
+// updateGatewayDevice writes to the device information to the gateway entity
+func updateGatewayDevice(ctx context.Context, deviceInfo *protos.GatewayDeviceInfo, hwID *protos.AccessGatewayID, challengeKey *protos.ChallengeKey) error {
+	networkID := deviceInfo.NetworkId
+	gatewayID := deviceInfo.LogicalId
+
+	ent, err := configurator.LoadEntity(
+		ctx,
+		networkID, orc8r.MagmadGatewayType, gatewayID,
+		configurator.EntityLoadCriteria{},
+		serdes.Entity,
+	)
+	if err != nil {
+		return err
+	}
+
+	device := createGatewayDevice(hwID, challengeKey)
+
+	status, err := wrappers.GetGatewayStatus(ctx, networkID, ent.PhysicalID)
+	if err != nil && err != merrors.ErrNotFound {
+		return err
+	}
+	gw := (&models.MagmadGateway{}).FromBackendModels(ent, device, status)
+
+	_, err = configurator.UpdateEntities(ctx, networkID, gw.ToEntityUpdateCriteria(ent), serdes.Entity)
+	if err != nil {
+		return err
+	}
```

Comment:
We should be able to do this with the "mutable" version of the gateway object -- i.e. without having to pull/integrate gateway state

Reply:
I was referencing how the update gateway handlers does it here https://github.com/christinewang5/magma/blob/cc9780ca2f8bef44d42a034ac744af07b7a445aa/orc8r/cloud/go/services/orchestrator/obsidian/handlers/gateway_handlers.go#L340-L362

But I was searching for the mutable version and it seems like there are only models for specific gateway types like cwf, lte, fed, etc. Lmk if there is a way to get the mutable version just for a generic gateway?

Reply:
Ah true weird that there isn't a mutable version of the magmad gateway. So let's still do it this way, except -- we can pass in the "status" as nil, since we don't care about that when we update the ent